File: lintcode/python/0606_kth_largest_element_II.py
#!/usr/bin/python -t

# heap
# 利用堆求第k大元素


import heapq

class Solution:
    """
    @param nums: an integer unsorted array
    @param k: an integer from 1 to n
    @return: the kth largest element
    """
    def kthLargestElement2(self, nums, k):
        # write your code here
        heap = []
        
        for num in nums:
            heapq.heappush(heap, num)
            if len(heap) > k:
                ret = heapq.heappop(heap)
                #print ret
                
        ret = heapq.heappop(heap)
            
        return ret
        
        

File: lintcode/golang/0606_kth_largest_element_II.go
//heap


import(
    "container/heap"
)

type IntHeap []int

func (h IntHeap) Len() int           { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x interface{}) {
	// Push and Pop use pointer receivers because they modify the slice's length,
	// not just its contents.
	*h = append(*h, x.(int))
}

func (h *IntHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

/**
 * @param nums: an integer unsorted array
 * @param k: an integer from 1 to n
 * @return: the kth largest element
 */
func kthLargestElement2 (nums []int, k int) int {
    // write your code here

    minHeap := &IntHeap{}

    for _, num := range nums {
        heap.Push(minHeap, num)
        if minHeap.Len() > k {
            heap.Pop(minHeap)
        }
    }

    ret := heap.Pop(minHeap).(int)

    return ret
}

